src/repr/src/adt/numeric.rs

```diff
@@ -588,3 +590,32 @@ pub fn rescale(n: &mut Numeric, scale: u8) -> Result<(), anyhow::Error> {
 
     Ok(())
 }
+
+/// Validates the typ_mod is valid for numeric type and returns a specified
+/// scale is appropriate (we ignore the specified precision).
+pub fn extract_typ_mod(typ_mod: &[u64]) -> Result<Option<u8>, anyhow::Error> {
+    let max_precision = NUMERIC_DATUM_MAX_PRECISION as u64;
+    util::validate_typ_mod(
+        "numeric",
+        &typ_mod,
+        &[("precision", 1, max_precision), ("scale", 0, max_precision)],
+    )?;
+
+    // Poor man's VecDeque
+    Ok(match typ_mod.len() {
+        0 | 1 => None,
+        2 => {
+            let precision = typ_mod[0] as u8;
+            let scale = typ_mod[1] as u8;
```

Comment:
Minor, and I know the code largely already exists, but not wild about the API of `validate_typ_mod` because it forces a separate `as` cast. Probably would be better to just `.try_into().unwrap_or_else(...)`, possibly with `validate_typ_mod` refactored into something that just generates the error messages consistently. (But hardcoding them in the three spots would probably feel fine too.)

Reply:
I took this in the opposite direction and genericized `validate_typ_mod` into also returning the `u64` values into the desired type (now called `extract_typ_mod`). This doesn't use `CastFrom` like you suggested, but does keep everything neat and in a single place, while avoiding any calls to unwrap or any extraneous casts. lmk if this is anathema to you and glad to reconsider the design.